crates/ruff_python_semantic/src/red_knot/types.rs

```diff
@@ -0,0 +1,705 @@
+        let db = setup_db();
+
+        db.memory_file_system().write_file("/src/a.py", "x = 10")?;
+        let a = system_path_to_file(&db, "/src/a.py").unwrap();
+
+        let parsed = parsed_module(&db, a);
+
+        let statement = parsed.suite().first().unwrap().as_assign_stmt().unwrap();
+
+        let literal_ty = expression_ty(&db, a, &statement.value);
+
+        assert_eq!(
+            format!("{}", literal_ty.display(&TypingContext::global(&db))),
+            "Literal[10]"
+        );
+
+        Ok(())
+    }
+
+    #[test]
+    fn dependency_public_symbol_type_change() -> anyhow::Result<()> {
+        let mut db = setup_db();
+
+        db.memory_file_system().write_files([
+            ("/src/a.py", "from foo import x"),
+            ("/src/foo.py", "x = 10\ndef foo(): ..."),
+        ])?;
+
+        let a = system_path_to_file(&db, "/src/a.py").unwrap();
+        let x_ty = public_symbol_ty_by_name(&db, a, "x").unwrap();
+
+        assert_eq!(
+            x_ty.display(&TypingContext::global(&db)).to_string(),
+            "Literal[10]"
+        );
+
+        // Change `x` to a different value
+        db.memory_file_system()
+            .write_file("/src/foo.py", "x = 20\ndef foo(): ...")?;
+
+        let foo = system_path_to_file(&db, "/src/foo.py").unwrap();
+        foo.touch(&mut db);
+        std::thread::sleep(Duration::from_millis(1));
+
+        let a = system_path_to_file(&db, "/src/a.py").unwrap();
+
+        db.clear_salsa_events();
+        let x_ty_2 = public_symbol_ty_by_name(&db, a, "x").unwrap();
+
+        assert_eq!(
+            x_ty_2.display(&TypingContext::global(&db)).to_string(),
+            "Literal[20]"
+        );
+
+        let a_root_scope = root_scope(&db, a);
+        let infer_types_ingredient_index = infer_types_ingredient_index(&db);
+        let events = db.take_sale_events();
+
+        assert!(events.iter().any(|event| {
+            if let salsa::EventKind::WillExecute { database_key } = &event.kind {
+                database_key.ingredient_index() == infer_types_ingredient_index
+                    && database_key.key_index() == a_root_scope.as_id()
+            } else {
+                false
+            }
+        }));
+
+        Ok(())
+    }
+
+    #[test]
+    fn dependency_non_public_symbol_change() -> anyhow::Result<()> {
+        let mut db = setup_db();
+
+        db.memory_file_system().write_files([
+            ("/src/a.py", "from foo import x"),
+            ("/src/foo.py", "x = 10\ndef foo(): y = 1"),
+        ])?;
+
+        let a = system_path_to_file(&db, "/src/a.py").unwrap();
+        let x_ty = public_symbol_ty_by_name(&db, a, "x").unwrap();
+
+        assert_eq!(
+            x_ty.display(&TypingContext::global(&db)).to_string(),
+            "Literal[10]"
+        );
+
+        db.memory_file_system()
+            .write_file("/src/foo.py", "x = 10\ndef foo(): pass")?;
+
+        let a = system_path_to_file(&db, "/src/a.py").unwrap();
```

Comment:
Not with Salsa 2022 but it won't compile with Salsa 3.0 if we don't because Salsa won't allow you to hold on to any Salsa values when calling a `&mut Db` function.